pkg/cache/v3/simple_test.go

```diff
@@ -375,3 +376,88 @@ func TestSnapshotClear(t *testing.T) {
 		t.Errorf("keys should be empty")
 	}
 }
+
+// BENCHMARKS =====================================================================================================
+
+func BenchmarkSnapshotCache(b *testing.B) {
+	c := cache.NewSnapshotCache(true, group{}, logger{t: &testing.T{}})
+
+	if _, err := c.GetSnapshot(key); err == nil {
+		b.Errorf("unexpected snapshot found for key %q", key)
+	}
+
+	if err := c.SetSnapshot(key, snapshot); err != nil {
+		b.Fatal(err)
+	}
+
+	snap, err := c.GetSnapshot(key)
+	if err != nil {
+		b.Fatal(err)
+	}
+	if !reflect.DeepEqual(snap, snapshot) {
+		b.Errorf("expect snapshot: %v, got: %v", snapshot, snap)
+	}
+
+	// try to get endpoints with incorrect list of names
+	// should not receive response
+	value, _ := c.CreateWatch(&discovery.DiscoveryRequest{TypeUrl: rsrc.EndpointType, ResourceNames: []string{"none"}})
+	select {
+	case out := <-value:
+		b.Errorf("watch for endpoints and mismatched names => got %v, want none", out)
+	case <-time.After(time.Second / 4):
+	}
+
+	for _, typ := range testTypes {
+		b.Run(typ, func(b *testing.B) {
+			for n := 0; n < b.N; n++ {
+				value, _ := c.CreateWatch(&discovery.DiscoveryRequest{TypeUrl: typ, ResourceNames: names[typ]})
+				select {
+				case out := <-value:
+					if gotVersion, _ := out.GetVersion(); gotVersion != version {
+						b.Errorf("got version %q, want %q", gotVersion, version)
+					}
+					if !reflect.DeepEqual(cache.IndexResourcesByName(out.(*cache.RawResponse).Resources), snapshot.GetResourcesAndTtl(typ)) {
+						b.Errorf("get resources %v, want %v", out.(*cache.RawResponse).Resources, snapshot.GetResourcesAndTtl(typ))
+					}
+				case <-time.After(time.Second):
+					b.Fatal("failed to receive snapshot response")
+				}
+			}
+		})
+	}
+}
```

Comment:
Would it be valuable to do one of these that execute the watch -> channel read concurrently in a goroutine? That way we'd get some insight into the lock contention impact

Reply:
Hmm interesting thought, I think that would be worth while adding, I'll address these comments then work on that change. Hopefully that provides some valuable insight

Reply:
Interesting, this code:
```go
func BenchmarkSnapshotCacheLockContention(b *testing.B) {
	c := cache.NewSnapshotCache(true, group{}, nil)

	if err := c.SetSnapshot(key, snapshot); err != nil {
		b.Fatal(err)
	}

	for _, typ := range testTypes {
		b.Run(typ, func(b *testing.B) {
			for n := 0; n < b.N; n++ {
				go func() {
					value, _ := c.CreateWatch(&discovery.DiscoveryRequest{TypeUrl: typ, ResourceNames: names[typ]})

					select {
					case <-value:
						// NO-OP since we don't want to eat the extra cycles in the benchmark
					case <-time.After(time.Second):
						b.Error("failed to receive snapshot response")
					}
				}()
			}
		})
	}
}
```
Yielded this:
```
BenchmarkSnapshotCacheLockContention/type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment-16         	  880964	      1302 ns/op	     977 B/op	      12 allocs/op
BenchmarkSnapshotCacheLockContention/type.googleapis.com/envoy.config.cluster.v3.Cluster-16                        	 1277925	       946.8 ns/op	     544 B/op	       8 allocs/op
BenchmarkSnapshotCacheLockContention/type.googleapis.com/envoy.config.route.v3.RouteConfiguration-16               	 1000000	      1341 ns/op	     960 B/op	      12 allocs/op
BenchmarkSnapshotCacheLockContention/type.googleapis.com/envoy.config.listener.v3.Listener-16                      	 1274304	       950.9 ns/op	     544 B/op	       8 allocs/op
BenchmarkSnapshotCacheLockContention/type.googleapis.com/envoy.service.runtime.v3.Runtime-16                       	 1285586	       962.9 ns/op	     544 B/op	       8 allocs/op
```

Maybe I misunderstood "watch -> channel read", but it seems like the lock contention isn't an issue?

Reply:
This just spawns the go routines and doesn't wait for the end of the test, you'd need to wait for the go routine to complete in order to capture a reasonable number. 

Thinking about it some more I'm not sure if the golang benchmark framework works well with this, since iirc it will try to run it with different Ns to get a better result, but I'm not sure if that approach works well if N is the concurrency of the test. Perhaps something where you have a fixed number of goroutines doing N operations + a WG to wait until the work is done would be a better measure?

Reply:
So my coworker @dougfort and myself were bashing our heads against this one, I ended up changing the function to something like this:
```go
func BenchmarkSnapshotCacheLockContention(b *testing.B) {
	c := cache.NewSnapshotCache(true, group{}, nil)

	if err := c.SetSnapshot(key, snapshot); err != nil {
		b.Fatal(err)
	}

	var wg sync.WaitGroup
	for i := 0; i < len(testTypes); i++ {
		wg.Add(1)
		go func() {
			for _, typ := range testTypes {
				b.Run(typ, func(b *testing.B) {
					for n := 0; n < b.N; n++ {
						value, _ := c.CreateWatch(&discovery.DiscoveryRequest{TypeUrl: typ, ResourceNames: names[typ]})

						select {
						case <-value:
							// NO-OP since we don't want to eat the extra cycles in the benchmark
						case <-time.After(time.Second):
							b.Error("failed to receive snapshot response")
						}
					}
				})
			}
			wg.Done()
		}()
	}

	// Hold the benchmark until all go routines have finished processing
	wg.Wait()
}
```
Which started that fixed number of go routine group, and then ran the benchmarks, but we found that if you call `b.Run()` inside of a go routine, the benchmarking framework fights for a global lock that results in a panic:
```
fatal error: sync: unlock of unlocked mutex
fatal error: sync: inconsistent mutex state

goroutine 169 [running]:
runtime.throw(0x17b65e3, 0x1e)
	/usr/local/Cellar/go/1.16.4/libexec/src/runtime/panic.go:1117 +0x72 fp=0xc000116638 sp=0xc000116608 pc=0x1039812
sync.throw(0x17b65e3, 0x1e)
	/usr/local/Cellar/go/1.16.4/libexec/src/runtime/panic.go:1103 +0x35 fp=0xc000116658 sp=0xc000116638 pc=0x106d955
sync.(*Mutex).unlockSlow(0x1d7d3c0, 0xffffffff)
	/usr/local/Cellar/go/1.16.4/libexec/src/sync/mutex.go:196 +0xd8 fp=0xc000116680 sp=0xc000116658 pc=0x107c158
sync.(*Mutex).Unlock(...)
	/usr/local/Cellar/go/1.16.4/libexec/src/sync/mutex.go:190
testing.(*B).Run(0xc000192d80, 0x17c7daa, 0x42, 0xc000722040, 0x13a228400)
	/usr/local/Cellar/go/1.16.4/libexec/src/testing/benchmark.go:618 +0x5a5 fp=0xc000116768 sp=0xc000116680 pc=0x110b2c5
github.com/envoyproxy/go-control-plane/pkg/cache/v3_test.BenchmarkSnapshotCacheLockContention.func1(0xc000192d80, 0x187a538, 0xc0078f00c0, 0xc000380000)
	/Users/alecholmes/Developer/alecholmez/go-control-plane/pkg/cache/v3/simple_test.go:434 +0x105 fp=0xc0001167c0 sp=0xc000116768 pc=0x15e97a5
runtime.goexit()
	/usr/local/Cellar/go/1.16.4/libexec/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc0001167c8 sp=0xc0001167c0 pc=0x1072a21
created by github.com/envoyproxy/go-control-plane/pkg/cache/v3_test.BenchmarkSnapshotCacheLockContention
	/Users/alecholmes/Developer/alecholmez/go-control-plane/pkg/cache/v3/simple_test.go:432 +0x1f1
```
Running this code with `RunParallel()` instead yields:
```
BenchmarkSnapshotCacheLockContention-16                                                                   	   46651	     26168 ns/op	   18055 B/op	     240 allocs/op
```
AFAIK though, this is not an accurate representation of the lock contention overhead. Honestly it might be worth scrapping this benchmark and moving forward with the full e2e test to really understand the hidden metrics we need. What do you think @snowp?

@dougfort if I missed anything feel free to add